src/main/java/net/coderbot/iris/mixin/vertices/MixinBufferBuilder.java

```diff
@@ -172,121 +142,47 @@ public void endBlock() {
-	private void computeTangents() {
-		// Capture all of the relevant vertex positions
-		float x0 = quad.x(0);
-		float y0 = quad.y(0);
-		float z0 = quad.z(0);
-
-		float x1 = quad.x(1);
-		float y1 = quad.y(1);
-		float z1 = quad.z(1);
-
-		float x2 = quad.x(2);
-		float y2 = quad.y(2);
-		float z2 = quad.z(2);
-
-		float edge1x = x1 - x0;
-		float edge1y = y1 - y0;
-		float edge1z = z1 - z0;
-
-		float edge2x = x2 - x0;
-		float edge2y = y2 - y0;
-		float edge2z = z2 - z0;
-
-		float u0 = quad.u(0);
-		float v0 = quad.v(0);
+	@Unique
+	private void fillExtendedData(int vertexAmount) {
+		vertexCount = 0;
 
-		float u1 = quad.u(1);
-		float v1 = quad.v(1);
+		// TODO: Keep this in sync with the extensions
+		int extendedDataLength = (2 * 2) + (2 * 4) + (1 * 4);
 
-		float u2 = quad.u(2);
-		float v2 = quad.v(2);
+		int stride = this.format.getVertexSize();
 
-		float deltaU1 = u1 - u0;
-		float deltaV1 = v1 - v0;
-		float deltaU2 = u2 - u0;
-		float deltaV2 = v2 - v0;
+		polygon.setup(buffer, this.nextElementByte, stride, vertexAmount);
 
-		float fdenom = deltaU1 * deltaV2 - deltaU2 * deltaV1;
-		float f;
+		float midU = 0;
+		float midV = 0;
 
-		if (fdenom == 0.0) {
-			f = 1.0f;
-		} else {
-			f = 1.0f / fdenom;
+		for (int vertex = 0; vertex < vertexAmount; vertex++) {
+			midU += polygon.u(vertex);
+			midV += polygon.v(vertex);
 		}
 
-		float tangentx = f * (deltaV2 * edge1x - deltaV1 * edge2x);
-		float tangenty = f * (deltaV2 * edge1y - deltaV1 * edge2y);
-		float tangentz = f * (deltaV2 * edge1z - deltaV1 * edge2z);
-		float tcoeff = rsqrt(tangentx * tangentx + tangenty * tangenty + tangentz * tangentz);
-		tangentx *= tcoeff;
-		tangenty *= tcoeff;
-		tangentz *= tcoeff;
-
-		float bitangentx = f * (-deltaU2 * edge1x + deltaU1 * edge2x);
-		float bitangenty = f * (-deltaU2 * edge1y + deltaU1 * edge2y);
-		float bitangentz = f * (-deltaU2 * edge1z + deltaU1 * edge2z);
-		float bitcoeff = rsqrt(bitangentx * bitangentx + bitangenty * bitangenty + bitangentz * bitangentz);
-		bitangentx *= bitcoeff;
-		bitangenty *= bitcoeff;
-		bitangentz *= bitcoeff;
-
-		// predicted bitangent = tangent × normal
-		// Compute the determinant of the following matrix to get the cross product
-		//  i  j  k
-		// tx ty tz
-		// nx ny nz
-
-		// Be very careful when writing out complex multi-step calculations
-		// such as vector cross products! The calculation for pbitangentz
-		// used to be broken because it multiplied values in the wrong order.
-
-		float pbitangentx =   tangenty * normal.z - tangentz * normal.y;
-		float pbitangenty = -(tangentx * normal.z - tangentz * normal.x);
-		float pbitangentz =   tangentx * normal.y - tangenty * normal.x;
-
-		float dot = (bitangentx * pbitangentx) + (bitangenty * pbitangenty) + (bitangentz * pbitangentz);
-		float tangentW;
-
-		if (dot < 0) {
-			tangentW = -1.0F;
+		midU /= vertexAmount;
+		midV /= vertexAmount;
+
+		if (vertexAmount == 3) {
+			NormalHelper.computeFaceNormal(normal, (TriView) polygon);
```

Comment:
I like most of the code here but the fact that this changes behavior just based on a seemingly unnecessary cast seems a bit too implicit for me. I don't want removing a seemingly unnecessary cast to change the behavior of the program.

Reply:
Possible solutions:
- Rename the method to `computeFaceNormalTri` or similar
- Split `BufferBuilderPolygonView` into `BufferBuilderTriView` and `BufferBuilderQuadView` and apply one of the following:
  - Allow the polygon itself or a separate handler to fill in the vertex data
  - Duplicate the `fillExtendedData` method (one for quads, one for tris)

Which solution should be applied?

Reply:
The rename approach seems to be the most straightforward for now.